fdbclient/include/fdbclient/KeyBackedTypes.actor.h

```diff
@@ -417,35 +423,44 @@ class KeyBackedProperty {
 		}
 	}
 
-	template <class Transaction>
-	void set(Transaction tr, T const& val) {
-		if constexpr (is_transaction_creator<Transaction>) {
-			return runTransaction(tr, [=, self = *this](decltype(tr->createTransaction()) tr) {
+	template <class TransactionContext>
+	std::enable_if_t<is_transaction_creator<TransactionContext>, Future<Void>> set(TransactionContext tcx,
+	                                                                               T const& val) {
+		return runTransaction(tcx, [this, val](decltype(tcx->createTransaction()) tr) {
+			if constexpr (SystemAccess) {
 				tr->setOption(FDBTransactionOptions::ACCESS_SYSTEM_KEYS);
-				tr->setOption(FDBTransactionOptions::LOCK_AWARE);
-				return self.set(tr, val);
-			});
-		} else {
-			tr->set(key, packValue(val));
-			if (trigger.present()) {
-				trigger->update(tr);
 			}
-		}
+			tr->setOption(FDBTransactionOptions::LOCK_AWARE);
+			set(tr, val);
+			return Future<Void>(Void());
+		});
 	}
 
 	template <class Transaction>
-	void clear(Transaction tr) {
-		if constexpr (is_transaction_creator<Transaction>) {
-			return runTransaction(tr, [=, self = *this](decltype(tr->createTransaction()) tr) {
+	std::enable_if_t<!is_transaction_creator<Transaction>, void> set(Transaction tr, T const& val) {
+		tr->set(key, packValue(val));
+		if (trigger.present()) {
+			trigger->update(tr);
+		}
+	}
+
+	template <class TransactionContext>
+	std::enable_if_t<is_transaction_creator<TransactionContext>, Future<Void>> clear(TransactionContext tcx) {
```

Comment:
Why `std::enable_if_t` instead of the former `if constexpr`? Is there a reason to avoid nested `if constexpr`'s ?

Reply:
The return types are different, so I needed to different definitions of the function

Reply:
I should add that this function was just broken before, but we apparently didn't call it with a database so we didn't notice.